numpy/f2py/cb_rules.py

```diff
@@ -33,114 +33,151 @@
     'cbtypedefs': 'typedef #rctype#(*#name#_typedef)(#optargs_td##args_td##strarglens_td##noargs#);',
     'body': """
 #begintitle#
-PyObject *#name#_capi = NULL;/*was Py_None*/
-PyTupleObject *#name#_args_capi = NULL;
-int #name#_nofargs = 0;
-jmp_buf #name#_jmpbuf;
+typedef struct {
+    PyObject *capi;
+    PyTupleObject *args_capi;
+    int nofargs;
+    jmp_buf jmpbuf;
+} #name#_t;
+
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && (__GNUC_MINOR__ >= 4))) && !defined(F2PY_USE_PYTHON_TLS)
+static __thread #name#_t *_active_#name# = NULL;
+static #name#_t *swap_active_#name#(#name#_t *ptr) {
+    #name#_t *prev = _active_#name#;
+    _active_#name# = ptr;
+    return prev;
+}
+static #name#_t *get_active_#name#() {
+    return _active_#name#;
+}
+#elif defined(_MSC_VER) && !defined(F2PY_USE_PYTHON_TLS)
+static __declspec(thread) #name#_t *_active_#name# = NULL;
+static #name#_t *swap_active_#name#(#name#_t *ptr) {
+    #name#_t *prev = _active_#name#;
+    _active_#name# = ptr;
+    return prev;
+}
+static #name#_t *get_active_#name#() {
```

Comment:
see above.

Reply:
Does that missing arguments need `void`?

Reply:
Yes, fixed (and fixed also the othe things).